src/code/sys_cmpdma.c

```diff
@@ -19,14 +19,152 @@
+        if ((sp8.a & 1) == 1) {
+            tempsp8 = sp8.ab;
+
+            spC.r = ((tempsp8 >> 0x1B) * 0xFF) / 31;
+            spC.g = (((tempsp8 << 5) >> 0x1B) * 0xFF) / 31;
+            spC.b = (((tempsp8 << 0xA) >> 0x1B) * 0xFF) / 31;
+            spC.a = 0xFF;
+
+        } else if (sp8.a == 0) {
+            spC.rgba = 0;
+
+        } else {
+            sp4.a = sp8.a;
+            sp8.a = *(phi_v1++);
+
+            tempsp4 = sp4.ab;
+            tempsp8 = sp8.ab;
+
+            spC.r = ((tempsp8 >> 0x1B) * 8) | (tempsp4 >> 0x1D);
+            spC.g = (((tempsp8 << 5) >> 0x1B) * 8) | ((tempsp4 * 8) >> 0x1D);
+            spC.b = (((tempsp8 << 0xA) >> 0x1B) * 8) | ((tempsp4 << 6) >> 0x1D);
+            tempsp8 = sp8.a;
+            spC.a = ((tempsp8 & 1) << 7) | ((((tempsp4 << 9) >> 0x1B) * 0x7F) / 63);
+        }
+
+        *(phi_v0++) = spC.rgba;
+    } while ((u32)phi_a0 < (u32)dst);
+}
+#else
+void func_80178AC0(u16* src, void* dst, u32 size);
 #pragma GLOBAL_ASM("asm/non_matchings/code/sys_cmpdma/func_80178AC0.s")
+#endif
+
+void func_80178C80(u8* vrom, s32 id, s32* start, s32* size, s32* flag) {
+    u32 dataStart;
+    u32 refOff;
+
+    DmaMgr_DmaRomToRam(vrom, (void*)&D_801FBBE0, sizeof(u32));
+
+    dataStart = D_801FBBE0.data1;
+    refOff = id * sizeof(u32);
+
+    // if id is >= idMax
+    if (refOff > (dataStart - 4)) {
+        *start = vrom;
+        *size = 0;
+    } else if (refOff == 0) {
+        // get offset start of next file, i.e. size of first file
+        DmaMgr_DmaRomToRam(vrom + 4, (void*)&D_801FBBE0, sizeof(u32));
+        *start = vrom + dataStart;
+        *size = D_801FBBE0.data1;
+    } else {
+        // get offset start, offset end from dataStart
+        DmaMgr_DmaRomToRam(refOff + vrom, (void*)&D_801FBBE0, sizeof(u32) * 2);
+        *start = D_801FBBE0.data1 + vrom + dataStart;
+        *size = D_801FBBE0.data2 - D_801FBBE0.data1;
+    }
+    *flag = 0;
+}
+
+void func_80178D7C(s32 romStart, s32 size, void* dst) {
+    if (size != 0) {
+        Yaz0_Decompress(romStart, dst, size);
+    }
+}
+
+void func_80178DAC(u32 vrom, s32 id, void* dst, s32 size) {
+    s32 romStart;
+    s32 compressedSize;
+    s32 flag;
+
+    func_80178C80(vrom, id, &romStart, &compressedSize, &flag);
+    if (flag & 1) {
+        void* temp = SystemArena_Malloc(0x1000);
+        func_80178D7C(romStart, compressedSize, temp);
+        func_80178AC0(temp, dst, size);
+        SystemArena_Free(temp);
+    } else {
+        func_80178D7C(romStart, compressedSize, dst);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_cmpdma/func_80178C80.s")
+void func_80178E3C(uintptr_t vrom, s32 id, void* dst, s32 size) {
+    func_80178DAC(DmaMgr_TranslateVromToRom(vrom), id, dst, size);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_cmpdma/func_80178D7C.s")
+void func_80178E7C(uintptr_t vrom, void* dst, s32 arg2) {
+    u32 temp_s2 = DmaMgr_TranslateVromToRom(vrom);
+    u32 i;
+    u32 end;
+    s32 phi_s1;
+    u32 var;
```

Comment:
```suggestion
    u32 dataStart;
```

Reply:
mentioned in the other comment, but I don't see the logic for making this dataStart

Reply:
never mind, this was actually correct; it's calculating the number of files in the segment